src/metrics.rs

```diff
@@ -0,0 +1,177 @@
+
+    if let Some(t) = input_val.get("time") {
+        output_val["time"] = t.clone();
+    }
+
+    if let Some(op_value) = input_val.get("fields").and_then(|f| f.get("operation")) {
+        match from_value(op_value.clone())? {
+            Operation::Snapshot {
+                repo,
+                op: Some(op),
+                error: _,
+                latency,
+            } => {
+                output_val["repo"] = Value::String(repo);
+                if let Some(latency) = Number::from_f64(latency as f64) {
+                    output_val["latency"] = Value::Number(latency);
+                }
+                output_val["dura_branch"] = Value::String(op.dura_branch);
+                output_val["commit_hash"] = Value::String(op.commit_hash);
+                output_val["base_hash"] = Value::String(op.base_hash);
+            }
+            _ => return Ok(None),
+        }
+    } else {
+        return Ok(None);
+    }
+
+    Ok(Some(output_val))
+}
+
+/// Use the info captured from scrape_log to open a repo and capture information about the commit
+///
+/// The repo_cache is retained between calls. This cache seems to cut runtime by 50% in a
+/// completely non-scientific measure. It still seems to take unexpectedly long, probably because
+/// it still has to open lots of files (for each commit & tree object) behind the scenes, and this
+/// is inherently not cache-able.
+fn scrape_git(
+    value: &mut Value,
+    repo_cache: &mut HashMap<String, Rc<Repository>>,
+) -> Result<(), git2::Error> {
+    if let Some(repo_path_value) = value.get("repo") {
+        let repo_path = repo_path_value.as_str().unwrap_or("");
+        let repo = match repo_cache.get(repo_path) {
+            Some(repo) => Rc::clone(repo),
+            None => {
+                let repo = Rc::new(Repository::open(repo_path)?);
+                repo_cache.insert(repo_path.to_string(), Rc::clone(&repo));
+                repo
+            }
+        };
+        let commit_opt = value
+            .get("commit_hash")
+            .and_then(|c| c.as_str())
+            .and_then(|c| Oid::from_str(c).ok())
+            .and_then(|c| repo.find_commit(c).ok());
+        let parent_commit = commit_opt.as_ref().and_then(|c| c.parents().last());
+        if let (Some(commit), Some(parent)) = (commit_opt, parent_commit) {
+            let diff =
+                repo.diff_tree_to_tree(Some(&parent.tree()?), Some(&commit.tree()?), None)?;
+            let stats = diff.stats()?;
+            value["num_files_changed"] = json!(stats.files_changed());
+            value["insertions"] = json!(stats.insertions());
+            value["deletions"] = json!(stats.deletions());
+
+            let files: Vec<_> = diff
+                .deltas()
+                .flat_map(|d| d.new_file().path())
+                .map(|p| p.to_str())
+                .collect();
+            value["files_changed"] = json!(files);
+        };
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Is there a reason these tests are inline rather than in their own file? Ideally they should be in the `tests` dir to be consistent

Reply:
because it's a private function. There's no reason to make it public, other than to test. I've generally been following the [official guidance on unit vs integration tests](https://doc.rust-lang.org/book/ch11-03-test-organization.html) although that division has been historically muddy. I've been calling any test that uses the file system an integration test, and making it external.